File: leetcode/problems/ts/maximum-depth-of-n-ary-tree.ts
// by @codeAbinash
// Time : O(n)
// Space : O(n)

class Node {
  val: number;
  children: Node[];
  constructor(val?: number, children?: Node[]) {
    this.val = val === undefined ? 0 : val;
    this.children = children === undefined ? [] : children;
  }
}

function maxDepth(root: Node | null): number {
  if (!root) return 0;
  let max = 0;
  for (let child of root.children) max = Math.max(max, maxDepth(child));
  return max + 1;
}

File: leetcode/problems/cpp/maximum-depth-of-n-ary-tree.cpp
// by @codeAbinash
// Time : O(n)
// Space : O(n)

#include "queue"
#include "vector"
using namespace std;

class Node {
  public:
   int val;
   vector<Node*> children;
   Node() {}
   Node(int _val) { val = _val; }
   Node(int _val, vector<Node*> _children) {
      val = _val;
      children = _children;
   }
};

// Using DFS
class Solution {
  public:
   int maxDepth(Node* root) {
      if (!root) return 0;
      int depth = 0;
      for (auto child : root->children) depth = max(depth, maxDepth(child));
      return 1 + depth;
   }
};

// using BFS
class Solution {
  public:
   int maxDepth(Node* root) {
      if (!root) return 0;
      int depth = 0;
      queue<Node*> q;
      q.push(root);

      while (!q.empty()) {
         depth++;
         int width = q.size();
         while (width--) {
            auto node = q.front();
            q.pop();
            for (auto child : node->children)
               if (child) q.push(child);
         }
      }
      return depth;
   }
};
